sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -351,7 +351,7 @@ private[sql] class HDFSBackedStateStoreProvider extends StateStoreProvider with
   }
 
   override def close(): Unit = {
-    synchronized { loadedMaps.values.asScala.foreach(_.clear()) }
+    synchronized { loadedMaps.keySet().asScala.toSeq.foreach(loadedMaps.remove) }
```

Comment:
Flipping the coin, we can also resolve this issue via explicitly copying the map when snapshotting. This way seems to be simpler to understand, but I see the proposal as more optimized so OK for me.

Reply:
Yea for large state - our memory usage would increase too. So thought that this way might be better - to continue working with the shallow references ?

Reply:
It may not be very huge as we are only copying the "map" and the memory usage of UnsafeRow would be the same, but the proposed way is still an optimized one (removing the overhead of copying) so I agree it's better.